File: frontend/src/lib/api/extraction.ts
/**
 * Typed API functions for structured-information extraction (Phase 14).
 *
 * Matches backend/app/schemas/extraction.py and backend/app/api/extractions.py.
 *
 * Endpoints consumed:
 *   POST /extractions                         → createExtractionRun()
 *   GET  /extractions/{extractionId}          → getExtractionRun()
 *   GET  /documents/{documentId}/extractions  → listExtractionRuns()
 */

import { get, post } from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export type ExtractionStatus = 'PENDING' | 'PROCESSING' | 'COMPLETED' | 'FAILED'
export type ExtractionCategory = 'requirement' | 'risk' | 'date' | 'party'

/** One extracted item with full citation-shaped provenance. */
export interface ExtractionItem {
  id: string
  category: ExtractionCategory
  item_index: number
  label: string
  detail: Record<string, unknown>
  chunk_id: string
  page_id: string
  page_number: number
  section: string | null
  quoted_text: string
  char_start: number | null
  char_end: number | null
  relevance_score: number | null
}

/** Items grouped by category — present only when status is COMPLETED. */
export interface ExtractionItemsByCategory {
  requirement: ExtractionItem[]
  risk: ExtractionItem[]
  date: ExtractionItem[]
  party: ExtractionItem[]
}

/** One extraction run (status + metadata). */
export interface ExtractionRun {
  id: string
  document_id: string
  document_version_id: string
  schema_key: string
  status: ExtractionStatus
  model: string | null
  prompt_version: string | null
  error_message: string | null
  created_at: string
  completed_at: string | null
}

/** Run detail — items included (grouped by category) when COMPLETED. */
export interface ExtractionRunDetail extends ExtractionRun {
  items: ExtractionItemsByCategory | null
}

/** Paginated run history for a document. */
export interface ExtractionRunsResponse {
  document_id: string
  total: number
  items: ExtractionRun[]
}

// ─── Request types ─────────────────────────────────────────────────────────────

export interface CreateExtractionRequest {
  document_id: string
  version?: number
  schema_key?: 'standard_v1'
}

// ─── API functions ─────────────────────────────────────────────────────────────

/**
 * Run structured extraction — always a NEW run (audit trail).  Requires the
 * `extraction:create` permission.  Poll `getExtractionRun()` until terminal.
 */
export async function createExtractionRun(
  body: CreateExtractionRequest,
): Promise<ExtractionRun> {
  return post<ExtractionRun>('/extractions', {
    schema_key: 'standard_v1',
    ...body,
  })
}

/**
 * Get one extraction run; items (grouped by category) are included when
 * status is COMPLETED.
 */
export async function getExtractionRun(
  extractionId: string,
): Promise<ExtractionRunDetail> {
  return get<ExtractionRunDetail>(`/extractions/${extractionId}`)
}

/**
 * Run history for a document, most recent first (auditable list).
 */
export async function listExtractionRuns(
  documentId: string,
  params?: { limit?: number; offset?: number },
): Promise<ExtractionRunsResponse> {
  const query = new URLSearchParams()
  if (params?.limit != null) query.set('limit', String(params.limit))
  if (params?.offset != null) query.set('offset', String(params.offset))
  const qs = query.toString()
  return get<ExtractionRunsResponse>(
    `/documents/${documentId}/extractions${qs ? `?${qs}` : ''}`,
  )
}

File: backend/app/schemas/extraction.py
"""
Pydantic v2 schemas for the structured-extraction API (Phase 14).

Request/response models for:
  - POST /extractions                          — create a new run
  - GET  /extractions/{extraction_id}          — run status + items
  - GET  /documents/{document_id}/extractions  — run history

All response models use from_attributes = True (ORM mode), following
schemas/comparison.py's exact convention.

See PHASE-14-IMPLEMENTATION-PLAN.md §5.13, §13.
"""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


class _OrmBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)


# ── Requests ──────────────────────────────────────────────────────────────────

class ExtractionCreateRequest(BaseModel):
    """Request body for POST /extractions."""

    document_id: str = Field(description="ID of the document to extract from.")
    version: Optional[int] = Field(
        default=None,
        description="Version number; null = the document's current version.",
    )
    schema_key: Literal["standard_v1"] = Field(
        default="standard_v1",
        description="Closed schema enum in V1 (bound anyway — never free text).",
    )


# ── Responses ─────────────────────────────────────────────────────────────────

class ExtractionItemResponse(BaseModel):
    """One extracted item with full citation-shaped provenance."""

    id: str
    category: str
    item_index: int
    label: str
    detail: dict[str, Any] = Field(default_factory=dict)
    chunk_id: str
    page_id: str
    page_number: int
    section: Optional[str] = None
    quoted_text: str
    char_start: Optional[int] = None
    char_end: Optional[int] = None
    relevance_score: Optional[float] = None


class ExtractionItemsByCategory(BaseModel):
    """Items grouped by category (only present when status is COMPLETED)."""

    requirement: list[ExtractionItemResponse] = Field(default_factory=list)
    risk: list[ExtractionItemResponse] = Field(default_factory=list)
    date: list[ExtractionItemResponse] = Field(default_factory=list)
    party: list[ExtractionItemResponse] = Field(default_factory=list)


class ExtractionRunResponse(_OrmBase):
    """One extraction run (status + metadata; items on the detail endpoint)."""

    id: str
    document_id: str
    document_version_id: str
    schema_key: str
    status: str
    model: Optional[str] = None
    prompt_version: Optional[str] = None
    error_message: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None


class ExtractionRunDetailResponse(ExtractionRunResponse):
    """Run detail — items included (grouped by category) when COMPLETED."""

    items: Optional[ExtractionItemsByCategory] = None


class ExtractionRunsResponse(BaseModel):
    """Paginated run history for a document (most recent first)."""

    document_id: str
    total: int
    items: list[ExtractionRunResponse]
